src/types/seat/drag_icon.rs

```diff
@@ -0,0 +1,162 @@
+use wlroots_sys::wlr_drag_icon;
+use std::cell::Cell;
+use std::rc::{Rc, Weak};
+use std::hash::{Hash, Hasher};
+use std::panic;
+use errors::{HandleErr, HandleResult};
+use DragIconHandler;
+
+use {SurfaceHandle};
+
+#[derive(Debug)]
+pub struct DragIcon {
+    liveliness: Rc<Cell<bool>>,
+    drag_icon: *mut wlr_drag_icon
+}
+
+impl DragIcon {
+    pub(crate) unsafe fn new(drag_icon: *mut wlr_drag_icon) -> Self {
+        let liveliness = Rc::new(Cell::new(false));
+        let state = Box::new(DragIconState {
+            handle: Rc::downgrade(&liveliness),
+            drag_icon
+        });
+        (*drag_icon).data = Box::into_raw(state) as *mut _;
+        DragIcon {
+            liveliness,
+            drag_icon
+        }
+    }
+
+    /// Get a handle for the surface associated with this drag icon
+    pub fn surface(&mut self) -> SurfaceHandle {
+        unsafe {
+            let surface = (*self.drag_icon).surface;
+			if surface.is_null() {
```

Comment:
More weird indentation?

Reply:
My editor is messing up and doing tabs sometimes for some reason.